content/post/051-new-ipfs-release-process.md

```diff
@@ -0,0 +1,209 @@
+---
+date: 2019-07-18
+url: 2019-07-18-ipfs-release-process
+tags: IPFS, go-ipfs, release process, 
+title: Improving the IPFS Release Process
+author: Steven Allen, Alan Shaw, David Dias, Molly Mackinlay
+---
+
+go-ipfs is introducing a new release cycle and process to ensure reliable releases.
+
+## Background
+
+The past three releases have had a few critical regressions, several of which
+would have been caught by better release testing and released faster with an
+improved release process.
+
+* 0.4.21 had two performance regressions in bitswap:
+  * a throughput regression that should have been caught by regression testing
+    (now tested) but almost certainly would have been noticed by downstream
+    users given a longer release-testing process.
+  * a CPU utilization regression that only shows up with >10000 peers. Something
+    that only really shows up in certain production systems under heavy load.
+* 0.4.20 had a regression where adding multiple independent files in the same
+  add command didn't work ([#6254](https://github.com/ipfs/go-ipfs/pull/6255)).
+  A regression test has since been added but this _also_ would have been caught
+  with better cross-application testing.
+* 0.4.19 had multiple regressions:
+  * a regression in the docker container (introduced by
+    [#6040](https://github.com/ipfs/go-ipfs/pull/6040)) that would have been
+    caught by testing the go-ipfs docker image in more production environments.
+  * a CPU utilization regression in bitswap only seen under very high load. This
+    would have been caught by testing under production loads.
+  * panics in the DHT and QUIC modules that only show up under heavy load.
+
+## Actions
+  
+To prevent this pattern from continuing, we've taken several actions:
+
+1. To address the stability issues, we've introduced a new release process that
+   involves extensively testing releases in a wide variety of production
+   environments.
+2. To address the issue of slow releases, we've introduced a 6 week release
+   cycle.
+3. To address the issue of slow bug fixes, we've switched to semver and
+   introduced patch releases. The first patch release will be 0.4.22 and the
+   next _feature_ release will be 0.5.0.
+   
+At the same time, we've introduced a feature release _freeze_ until we can
+finish building out our testing infrastructure (a topic for another blog post).
+
+XXX: Should we write that one first? It's awkward to say "we're introducing a
```

Comment:
?

Reply:
I don't think the plan deserves a blog post at this time (maybe when fully implemented), but it **does** deserve more explanation than this. Maybe:

"....finish building out our testing infrastructure designed to evaluate IPFS and libp2p at scale - from single node performance tests to overall network health and efficiency under simulated usage conditions. This test infra should improve our confidence that every release improves upon the previous one, and empower IPFS devs to keep our builds green through tight integration with our software engineering pipeline."

Reply:
Reworked.